cpp/src/ast/linearizer.cpp

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/ast/linearizer.hpp>
+#include <cudf/ast/operators.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+
+#include <algorithm>
+#include <functional>
+#include <iterator>
+
+namespace cudf {
+
+namespace ast {
+
+namespace detail {
+
+cudf::size_type intermediate_counter::take()
+{
+  auto const first_missing =
+    this->used_values.size() ? this->find_first_missing(0, this->used_values.size() - 1) : 0;
+  this->used_values.insert(this->used_values.cbegin() + first_missing, first_missing);
+  this->max_used = std::max(max_used, first_missing + 1);
+  return first_missing;
+}
+
+void intermediate_counter::give(cudf::size_type value)
+{
+  auto const found = std::find(this->used_values.cbegin(), this->used_values.cend(), value);
+  if (found != this->used_values.cend()) { this->used_values.erase(found); }
+}
+
+cudf::size_type intermediate_counter::find_first_missing(cudf::size_type start,
+                                                         cudf::size_type end) const
+{
+  // Given a sorted container, find the smallest value not already in the container
+  if (start > end) return end + 1;
+
+  // If the value at an index is not equal to its index, it must be the missing value
+  if (start != used_values.at(start)) return start;
+
+  auto const mid = (start + end) / 2;
```

Comment:
Do you need to implement binary search from scratch? I admit I don't fully understand what you are doing. Documentation for these functions would help.